src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc

```diff
@@ -0,0 +1,279 @@
+  }
+
+  Unit PreVisit(const Attribute& attrib) override {
+    if (pre_visits_) nodes_.push_back(attrib.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const Attribute& attrib, Unit result) override {
+    if (post_visits_) nodes_.push_back(attrib.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const CanActAs& canActAs) override {
+    if (pre_visits_) nodes_.push_back(canActAs.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const CanActAs& canActAs, Unit result) override {
+    if (post_visits_) nodes_.push_back(canActAs.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const BaseFact& baseFact) override {
+    if (pre_visits_) nodes_.push_back(baseFact.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const BaseFact& baseFact, Unit result) override {
+    if (post_visits_) nodes_.push_back(baseFact.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const Fact& fact) override {
+    if (pre_visits_) nodes_.push_back(fact.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const Fact& fact, Unit result) override {
+    if (post_visits_) nodes_.push_back(fact.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const ConditionalAssertion& condAssertion) override {
+    if (pre_visits_) nodes_.push_back(condAssertion.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const ConditionalAssertion& condAssertion,
+                 Unit result) override {
+    if (post_visits_) nodes_.push_back(condAssertion.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const Assertion& assertion) override {
+    if (pre_visits_) nodes_.push_back(assertion.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const Assertion& assertion, Unit result) override {
+    if (post_visits_) nodes_.push_back(assertion.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const SaysAssertion& saysAssertion) override {
+    if (pre_visits_) nodes_.push_back(saysAssertion.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const SaysAssertion& saysAssertion, Unit result) override {
+    if (post_visits_) nodes_.push_back(saysAssertion.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const Query& query) override {
+    if (pre_visits_) nodes_.push_back(query.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const Query& query, Unit result) override {
+    if (post_visits_) nodes_.push_back(query.DebugPrint());
+    return result;
+  }
+
+  Unit PreVisit(const Program& program) override {
+    if (pre_visits_) nodes_.push_back(program.DebugPrint());
+    return Unit();
+  }
+  Unit PostVisit(const Program& program, Unit result) override {
+    if (post_visits_) nodes_.push_back(program.DebugPrint());
+    return result;
+  }
+
+  const std::vector<std::string>& nodes() const { return nodes_; }
+
+ private:
+  bool pre_visits_;
+  bool post_visits_;
+  std::vector<std::string> nodes_;
+};
+
+TEST(AuthLogicAstTraversingVisitorTest, SimpleTraversalTest) {
```

Comment:
It might be more straightforward with a gmock, but it is fine for now. 

The only question is if there is a way to break this down into smaller tests.

Reply:
Opened an issue just because there is another PR depending on this one (and IDK or don't remember what a gmock is yet) #666